### PR TITLE
Realsense reset fix when spawning multiple realsense cameras

### PR DESCRIPTION
**Pull request by abhijitmajumdar**
TBD

### CODE REVIEW

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 1]
```diff
@@ -158,10 +144,25 @@ void RealSenseNodeFactory::onInit()
...
 			privateNh.param("initial_reset", _initial_reset, false);
 			_device = getDevice();
+			if (_device && _initial_reset)
+			{
+				_initial_reset = false;
+				try
+				{
+					ROS_INFO("Resetting device...");
+					_device.hardware_reset();
+					std::this_thread::sleep_for (std::chrono::seconds(10));
```
Comment (130s) [inline realsense2_camera/src/realsense_node_factory.cpp:156]: Is this `seconds(10)` the time that the entire Realsense node halts for? If so I'm afraid I can imagine there are applications where this change may not be desirable.

I'm not sure though, a few recommendations:
- Make it configurable (from `launch` etc.)
- If we know minimum length of period that the driver safely completes resetting with, we can set that as a default.

Comment (shaun-edwards) [reply realsense2_camera/src/realsense_node_factory.cpp:156]: Agree with @130s.  I would make this a launch parameter (can be done in the future).

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 2]
```diff
@@ -158,10 +144,25 @@ void RealSenseNodeFactory::onInit()
...
-			std::function<void(rs2::event_information&)> change_device_callback_function = [this](rs2::event_information& info){change_device_callback(info);};
-			_ctx.set_devices_changed_callback(change_device_callback_function);
 			privateNh.param("initial_reset", _initial_reset, false);
 			_device = getDevice();
+			if (_device && _initial_reset)
+			{
+				_initial_reset = false;
+				try
+				{
+					ROS_INFO("Resetting device...");
```
Comment (shaun-edwards) [inline realsense2_camera/src/realsense_node_factory.cpp]: If you know the device SN, it might be helpful to include it in the INFO message.

Comment (abhijitmajumdar) [reply realsense2_camera/src/realsense_node_factory.cpp]: The driver does print that out as a INFO message in the rest of the code, during initialization

Comment (shaun-edwards) [reply realsense2_camera/src/realsense_node_factory.cpp]: So you will add it or no?

Comment (abhijitmajumdar) [reply realsense2_camera/src/realsense_node_factory.cpp]: ~Doesn't it become redundant then?~
Nevermind, I understood the suggestion wrong, will add the SN in the INFO

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 1 above)
Comment (abhijitmajumdar) [reply realsense2_camera/src/realsense_node_factory.cpp:156]: Instead, would it be advisable that I check let's say every second after the reset to see if its finished?

The upstream implementation of this is using interrupts, but there is a problem with that, where only 1 camera is detected after a reset, inside the interrupt routine

Comment (130s) [reply realsense2_camera/src/realsense_node_factory.cpp:156]: I guess user should have a control over the recovery time.

> would it be advisable that I check let's say every second after the reset to see if its finished?

I think that's nice if it works (does the reset function return something reliably?). In this case the user would be able to pass the maximum trial time etc.?

Comment (abhijitmajumdar) [reply realsense2_camera/src/realsense_node_factory.cpp:156]: Good point! I would have to look if the reset function returns something useful. If not I think we can still have the maximum wait time configurable by the user, while internally it resets and keeps checking every second, until it either finds the reconnected realsense, or timesout

Comment (130s) [reply realsense2_camera/src/realsense_node_factory.cpp:156]: That can be addressed later. We'll track the enhancement.

**Comment by 130s** [comment]
Internal testing confirmed this is a good addition. 

CI failure is tracked at https://github.com/plusone-robotics/realsense/issues/22

Merging.

**Pull request merged**